bsg_mem/bsg_mem_dma.cpp

```diff
@@ -1,59 +1,63 @@
 #include <cstdio>
 #include <cassert>
 #include <map>
-#include "bsg_test_dram_channel.hpp"
+#include "bsg_mem_dma.hpp"
 
 #ifdef DEBUG
 #define pr_dbg(fmt, ...)                        \
-    printf("[bsg_test_dram_channel]: " fmt, ##__VA_ARGS__)
+    do { printf("[bsg_mem_dma]: " fmt, ##__VA_ARGS__); fflush(stdout); } while (0)
 #else
 #define pr_dbg(fmt, ...)
 #endif
 
 using namespace std;
 
-namespace bsg_test_dram_channel {
+namespace bsg_mem_dma {
     std::map<parameter_t, Memory *> global_memories;
 }
 
-using namespace bsg_test_dram_channel;
+using namespace bsg_mem_dma;
 
-extern "C" void * bsg_test_dram_channel_init(
+extern "C" void * bsg_mem_dma_init(
     parameter_t id,
     parameter_t channel_addr_width_p,
     parameter_t data_width_p,
     parameter_t mem_els_p,
     parameter_t init_mem_p
     )
 {
+    pr_dbg("id = %llu, addr_width_p=%llu, data_width_p=%llu, mem_els_p=%llu\n",
+           id, channel_addr_width_p, data_width_p, mem_els_p);
+    
     assert(data_width_p % 8 == 0);
-    Memory *memory =  new Memory(channel_addr_width_p, data_width_p, mem_els_p, init_mem_p);
+    Memory *memory =  new Memory(channel_addr_width_p, data_width_p, mem_els_p, init_mem_p, id);
     global_memories[id] = memory;
     return memory;
 }
 
-extern "C" byte_t bsg_test_dram_channel_get(
+extern "C" byte_t bsg_mem_dma_get(
     void *handle,
     address_t addr
     )
 {
     Memory *memory = reinterpret_cast<Memory*>(handle);
-    pr_dbg("getting 0x%08llx   (%02x)\n", addr, memory->get(addr));
+    pr_dbg("id = %llu: getting 0x%08llx   (%02x)\n", memory->_id, addr, memory->get(addr));
     return memory->get(addr);
 }
 
-extern "C" void bsg_test_dram_channel_set(
+extern "C" void bsg_mem_dma_set(
     void *handle,
```

Comment:
Why is it necessary to have this handle reference saved in verilog? Why not just have map<int,vector<byte>> as a static variable, and the module passes in its ID into these DPI calls?

Reply:
Saving the pointer means you don't need to do a lookup every time you make the DPI calls.